clash-prelude/src/Clash/Class/Num.hs

```diff
@@ -65,16 +63,16 @@ class (Bounded a, Num a) => SaturatingNum a where
   satSub  :: SaturationMode -> a -> a -> a
   -- | Multiplication with parameterizable over- and underflow behavior
   satMul :: SaturationMode -> a -> a -> a
-
--- | Get successor of (or in other words, add 1 to) given number
-satSucc :: SaturatingNum a => SaturationMode -> a -> a
-satSucc s n = satAdd s n 1
-{-# INLINE satSucc #-}
-
--- | Get predecessor of (or in other words, subtract 1 from) given number
-satPred :: SaturatingNum a => SaturationMode -> a -> a
-satPred s n = satSub s n 1
-{-# INLINE satPred #-}
+  -- | Get successor of (or in other words, add 1 to) given number
+  satSucc :: SaturationMode -> a -> a
+  -- Default method suitable for types that can represent the number 1
+  satSucc s n = satAdd s n 1
+  {-# INLINE satSucc #-}
+  -- | Get predecessor of (or in other words, subtract 1 from) given number
+  satPred :: SaturationMode -> a -> a
+  -- Default method suitable for types that can represent the number 1
+  satPred s n = satSub s n 1
```

Comment:
How about:

```haskell
  -- Instances should override this default implementation if their interval
  -- can be entirely negative. I.e., if they can be instantiated such that
  -- both 'minBound' and 'maxBound' are less than zero.
  satPred s n = 
    if maxBound @a > 0 then
      satSub s n 1
    else
      0
```

and

```haskell
  -- Instances should override this default implementation if their interval
  -- can be entirely negative. I.e., if they can be instantiated such that
  -- both 'minBound' and 'maxBound' are less than zero.
  satSucc s n = 
    if maxBound @a > 0 then
      satSucc s n 1
    else
      0
```

Reply:
This would remove the need for a custom implementation for all our types AFAIK.

Reply:
What about `SFixed 0 n` or `SFixed 1 n`? Unfortunately, it's not that simple.

Reply:
You're right!